doc/reference/trd-time.md

```diff
@@ -28,148 +29,342 @@ broad categories: alarms and timers. Alarms continuously increment a clock and
+    // rounding down any fractions.
+    fn ticks_from_seconds(s: u32) -> Self::Ticks;
 
-    fn is_armed(&self) -> bool;
+    // Returns the number of ticks in the provided number of milliseconds,
+    // rounding down any fractions.
+    fn ticks_from_ms(ms: u32) -> Self::Ticks;
+
+    // Returns the number of ticks in the provided number of microseconds,
+    // rounding down any fractions.
+    fn ticks_from_us(us: u32) -> Self::Ticks;
 }
 ```
 
 Frequency is defined with an [associated type] of the `Time` trait
-(`Time::Frequencey`). It MUST implement the `Frequency` trait, which has a
-single method, `frequency`. `frequency` returns the frequency in Hz, e.g. 1MHz
-is 1000000. Clients can use this to write clock-independent code. For example,
-to convert from seconds to clock tics:
+(`Time::Frequencey`). It MUST implement the `Frequency` trait, which
+has a single method, `frequency`. `frequency` returns the frequency in
+Hz, e.g. 1MHz is 1000000. Clients can use this to write code that is
+independent of the underlying frequency.
+
+An instance of `Time` or derived trait MUST NOT have a `Frequency`
+which is greater than its underlying frequency precision.  It must be
+able to accurately return every possible value in the range of `Ticks`
+without further quantization. It is therefore not allowed to take a
+32kHz clock and present it as an instance of `Time` with a frequency
+of `Freq16MHz`.
+
+`Frequency` allows a user of `Time` to know the granularity of ticks
+and so avoid quantization error when two different times map to the
+same time tick. For example, if a user of `Time` needs microsecond
+precision, then the associated type can be used to statically check
+that is is not put on top of an implementation with 32kHz precision.
+
+The three `ticks_from` methods are helper functions to convert values
+in seconds, milliseconds, or microseconds to a number of ticks. These
+three methods all round down the result. This means, for example, that
+if the `Time` instance has a frequency of 32kHz, calling
+`ticks_from_us(20)` returns 0, because a single tick of a 32kHz clock
+is 30.5 microseconds.
+
+[associated type]: https://doc.rust-lang.org/book/associated-types.html
+
+
+3 `Counter` and `OverflowClient` traits
+===============================
+
+The `Counter` trait is the abstraction of a free-running counter that
+can be started and stopped. This trait derives from the `Time` trait, so
+it has associated `Frequency` and `Tick` types. The `Counter` trait
+allows a client to register for callbacks when the counter overflows.
 
 ```rust
-fn seconds_to_tics<T: Time>(seconds: u32) -> u32 {
-    seconds * <T::Frequency>::frequency()
+pub trait OverflowClient {
+  fn overflow(&self);
+}
+
+pub trait Counter<'a>: Time {
+  fn start(&self) -> ReturnCode;
+  fn stop(&self) -> ReturnCode;
+  fn reset(&self) -> ReturnCode;
+  fn is_running(&self) -> bool;
+  fn set_overflow_client(&'a self, &'a dyn OverflowClient);
```

Comment:
Presumably, `OverflowClient::overflow` would be called in an interrupt handler or a deferred callback, as that is the normal Tock practice for callbacks. If so, it is possible that the type implementing `OverflowClient` calls `Time::now` between when the overflow has occurred and when the callback is called.

In other words, the following sequence of events is possible (assumes an 8-bit counter):

1. Client code calls `Time::now`, which returns 250.
2. An overflow happens, marking an interrupt as pending.
3. Client code calls `Time::now`, which returns 12.
4. The main event loop runs, invoking the interrupt handler.
5. The Counter calls `OverflowClient::overflow`, notifying the client of the overflow.

This seems confusing, and it appears it would be difficult to handle correctly. Worse, it's unclear how this sequence interacts with `stop` and`reset` (what if `reset` is called between steps 2 and 5?).

I recognize that this design may be the best solution to a difficult problem, but if so there needs to be documentation of how these methods interact with delayed overflows.

Reply:
Yes, this is a good point to mention in the document. It's a basic fact of live with deferred callbacks.

Reply:
Addressed.